src/XRP/network-clients/web-socket-network-client.ts

```diff
@@ -0,0 +1,155 @@
+import WebSocket = require('ws')
+import { XrpError, XrpErrorType } from '../shared'
+import {
+  WebSocketRequestOptions,
+  WebSocketResponse,
+  WebSocketStatusResponse,
+  WebSocketTransactionResponse,
+} from '../shared/rippled-web-socket-schema'
+
+const sleep = (ms: number): Promise<void> => {
+  return new Promise((resolve) => setTimeout(resolve, ms))
+}
+
+/**
+ * A network client for interacting with the rippled WebSocket API.
+ * @see https://xrpl.org/rippled-api.html
+ */
+export default class WebSocketNetworkClient {
+  private readonly socket: WebSocket
+  private accountCallbacks: Map<
+    string,
+    (data: WebSocketTransactionResponse) => void
+  >
+  private messageCallbacks: Map<string, (data: WebSocketResponse) => void>
+  private waiting: Map<string, WebSocketResponse | undefined>
+
+  /**
+   * Create a new WebSocketNetworkClient.
+   *
+   * @param webSocketUrl The URL of the rippled node to query.
+   * @see https://xrpl.org/monitor-incoming-payments-with-websocket.html
+   */
+  public constructor(
+    webSocketUrl: string,
+    private handleErrorMessage: (message: string) => void,
+  ) {
+    this.socket = new WebSocket(webSocketUrl)
+    this.accountCallbacks = new Map()
+    this.waiting = new Map()
+
+    this.messageCallbacks = new Map()
+    this.messageCallbacks.set('response', (data: WebSocketResponse) => {
+      const dataStatusResponse = data as WebSocketStatusResponse
+      this.waiting.set(dataStatusResponse.id, data)
+    })
+    this.messageCallbacks.set('transaction', this.handleTransaction)
+
+    this.socket.addEventListener('message', (event) => {
+      const parsedData = JSON.parse(event.data) as WebSocketResponse
+      const dataType = parsedData.type
+      const callback = this.messageCallbacks.get(dataType)
+      if (callback) {
+        callback(parsedData)
+      } else {
+        this.handleErrorMessage(
+          'Received unhandlable message: ' + (event.data as string),
+        )
+      }
+    })
+
+    this.socket.addEventListener('close', (event) => {
+      this.handleErrorMessage(
+        'Web socket disconnected, ' + (event.reason as string),
+      )
+    })
+
+    this.socket.addEventListener('error', (event) => {
+      this.handleErrorMessage('Error: ' + (event as string))
+    })
+  }
+
+  /**
+   * Properly handles incoming transactions from the web socket.
+   *
+   * @param data The web socket response received from the web socket.
+   */
+  private handleTransaction = (data: WebSocketResponse) => {
```

Comment:
can we make this type be a `WebSocketTransactionResponse` and let the caller figure out the type? that should keep the type checking in `handleTransaction` tighter

Reply:
The only caller of this function is the `message` event listener, which doesn't know anything specific about the handler and just knows it takes a `WebSocketResponse` object, so that won't work.

Reply:
well, it's only called to handle the `transaction` message type, right? you can enforce the type there by doing

```
this.messageCallbacks.set(
  'transaction', 
  (data: WebSocketResponse) => this.handleTransaction(data as WebSocketTransactionResponse),
)
```